### ISSUE TITLE
live streams not working on both twitch and youtube

### ISSUE THREAD

**Issue by Littlemac123** [ttalvitie/browservice#77]
I would love this to be fixed so i can watch my fav loli v-tubers on old computers <3
![Screenshot_2022-05-12_14-30-15](https://user-images.githubusercontent.com/39867553/168153846-d2888530-daf1-4d59-96ff-f184a14bccdb.png)

**Comment by ttalvitie** [comment]
At least with the nintendoloop stream, the issue seems to be that it requires proprietary codecs that are not part of the standard CEF build. I just updated the relevant part in the main README:

- By default, Browservice can't play videos that use proprietary audio/video codecs such as H264 and AAC, as the prebuilt CEF distribution [provided by Spotify](https://cef-builds.spotifycdn.com/index.html) does not include them. To add the codecs, build the CEF distribution by following the [instructions](https://bitbucket.org/chromiumembedded/cef/wiki/AutomatedBuildSetup.md) with the options `proprietary_codecs=true ffmpeg_branding=Chrome` appended to the environment variable `GN_DEFINES` (on Windows, you need to build the patched CEF version as shown in the [Windows build instructions](../blob/master/winbuild/README.md)). After this, you should proceed with the Browservice build process, using your self-built CEF instead of the prebuilt CEF distribution (in a [Linux build](../blob/master/BUILD.md), prior to running `setup_cef.sh`, copy the CEF distribution produced by the build to `cef.tar.bz2` instead of using `download_cef.sh` to download it). Note that building CEF takes a lot of time, memory and disk space. Also note that you may have to pay license fees to use the proprietary codecs legally, as they are encumbered by patents.

So in short, it is technically possible to get proprietary codecs working, but it is legally difficult or expensive because of how the proprietary codec licensing works. If you can take care of the legal side, the instructions above should help you create a custom build of Browservice with proprietary codec support. Apart from this, I can't really help with proprietary codecs.

**Comment by ttalvitie** [comment]
Yes, all twitch streams I picked pseudo-randomly (from both `m.twitch.tv` and `twitch.tv`) seem to require proprietary codecs.

**Comment by shuuryou** [comment]
Sorry to jump in out of the blue but I think the gold standard way of implementing streaming video and audio would be:

1. Check if website is supported by youtube-dl, if not stop
1. If supported, enable a toolbar button that opens a pop-up window with a Windows Media Player control (which they released of course for IE browsers but also as a Netscape plug-in)
1. Point the Windows Media Player to transcoding proxy; e.g. play the output of youtube-dl piped through ffmpeg (basically what https://github.com/atauenis/webone/blob/master/LinuxAndMacOS/yt.sh does)
1. When navigating away, destroy/close the pop-up if opened
1. Add a global setting to disable CEF playing video and audio if the user wishes (useful?)

This approach could fix many things:
* CEF not supporting proprietary media formats = no longer a problem
* Audio playing on proxy host computer = no longer a problem
* Lag due to the way rendering is implemented = no longer a problem

How it was done back in the golden days:
```html
<OBJECT ID="MediaPlayer" WIDTH="192" HEIGHT="190" CLASSID="CLSID:22D6F312-B0F6-11D0-94AB-0080C74C7E95"
STANDBY="Loading Windows Media Player components..." TYPE="application/x-oleobject">
<PARAM NAME="FileName" VALUE="videofilename.wmv">
<PARAM name="autostart" VALUE="false">
<PARAM name="ShowControls" VALUE="true">
<param name="ShowStatusBar" value="false">
<PARAM name="ShowDisplay" VALUE="false">
<EMBED TYPE="application/x-mplayer2" SRC="videofilename.wmv" NAME="MediaPlayer"
WIDTH="192" HEIGHT="190" ShowControls="1" ShowStatusBar="0" ShowDisplay="0" autostart="0"> </EMBED>
</OBJECT>
```

The HTML fragment for the `<object>` or `<embed>` tag that is eventually used on the client should be a HTML file on disk of the server so that clever people can easily change the `CLASSID` or `TYPE` to RealPlayer, VLC, Flash/FLV, or something Unix/MacOS specific, etc. and tweak the shell script doing the transcoding.

The list of sites supported by youtube-dl is quite extensive, to say the least:
https://ytdl-org.github.io/youtube-dl/supportedsites.html

These websites would most likely all just work if implemented like this. You could of course go overkill and determine at what point (x,y) and dimensions (w,h) is e.g. the `<video>` element and then overlay a `<object>` tag with CSS z-index 9999 on top of it, but then you need to handle scrolling it off the screen locally, etc. too. A toolbar button that opens a pop-up window locally is probably a rather quick addition?
